Address memory issues in `combine`

b-reyes opened the pull request:
This PR addresses issue #766 by allowing `combine` to directly write to a user provided zarr store in a distributed fashion. This is done as follows:

1. Perform a delayed `to_zarr` call, which produces all metadata (groups, chunks, etc.) for the final combined `EchoData` object
2. Directly write each `EchoData` object's group to the final zarr store in parallel using `to_zarr` and `region` 


Attention! This PR is currently a draft. A few things to note: 
- There exist other files in the `echodata` folder pertaining to combine. Several of these are prior attempts at the combine function and should be ignored. In this draft state, one should only look at the file named `zarr_combine.py` this is the file that implements the above items. 
- As of right now, I have not integrated this new combine functionality into the api, this will be done later.

b-reyes commented:
At the request of @leewujung, here is a list of TODOs for this PR (some may not be included, but I will try to update this list as I remember them): 

# TODOs
- [x] Figure out RuntimeError: error during blosc decompression: -1
    - This seems related to https://github.com/pangeo-data/pangeo/issues/196
    - I was unable to find the true underlying reason for this issue. 
    I tried several solutions that were suggested and none of them worked. 
    I think the reason for this error is that two processes are attempting 
    to access the same chunk file. This is unavoidable because zarr requires 
    chunks to be uniform (dask doesn't have this requirement).
    - setting retries=1 in compute seem to solve this issue. In my opinion this 
    is ok since we are writting in slices. Thus, if data is corrupted then the 
    retry will just write over the corrupted data (if successful)
- [ ] make sure it works for s3 bucket writing (we can use minio)
    - allow for `fs_map` input (make this a new issue/PR maybe)
- [x] run a selection of files from `noaa-wcsd-pds/data/raw/Bell_M._Shimada/SH1707/EK80`
- [ ]  run a selection of files from `noaa-wcsd-pds/data/raw/Bell_M._Shimada/SH2106/EK80`
    - These runs triggered an error in the current PR code because the `channel` dimension is not always in ascending order. This is currently being corrected in issue #815.
- [ ]  run a selection of files from `noaa-wcsd-pds/data/raw/Bell_M._Shimada/SH1701/EK60`
- [ ]  run a selection of files from OOI associated with the AZFP echosounder 
- [x] add attributes to each `EchoData` group
- [ ] create a check that all of the channels are the same
- [ ] check for and correct reversed time
- [ ] store old times if they needed to be corrected
- [ ] make sure that the combined files have ascending times  
- [x] change lazy naming for class
- [ ] Make sure the combine function works with different sized `range_sample` dimensions amongst the sensors

A couple of additional items to note:
- The Provenance attribute variables are named differently in this PR
- We introduce a routine that ensures all `ds_lists` have identical attributes. Runtime errors are raised if: 
    - The keys are not the same
    - The values are not identical
    - The keys `date_created`, `conversion_time` do not have the same types
- If attribute values are numpy arrays, then they will not be included in the `Provenance` group. Instead, these values will only appear in the attributes of the combined `EchoData` object.

emiliom commented:
Regarding our discussion today about the use of small batches of test data files to test `combine_echodata`: the existing tests in `test_echodata_combine.py` already define small batches of ek60, ek80 and azfp files that get combined. Those files are already in the Google Drive test data folder, and all the tests run successfully with the old/current `combine_echodata`. So, it seems like there's no need to download any additional files for testing. All those files are already available via the docker container test http server, and I believe they're also available via the minio server.

The tests would need to be modified so they also read from http and s3. But no new data files or live access to the external data files is needed, as far as I can tell -- unless you intend to test for characteristics not captured in those data files.

b-reyes commented:
I have just implemented the checking and storing of reversed time coordinates. After some discussion with @leewujung, we have arrived at a different naming scheme for the storage of the old time coordinates (if they are encountered), in comparison to how things were done in the old `combine.py`.

 If reversed times are found they will be corrected in the `EchoData` object (as they previously were), then we will store the old times (uncorrected times) in the `Provenance` group with the name: `group_old_timename` and dimension: `group_old_timename_dim`, where `group` and `timename` will be replaced with the appropriate `EchoData` group and time, respectively. Note that we do not want to set the dimension as a coordinate as this can cause an increase in memory usage when working with the lazy loaded version (since xarray loads in coordinates). 

Additionally, this old time will now have the attribute `comment:Uncorrected time from the combined group Group`, with `time` and `Group` appropriately set and all attributes that were previously associated with the variable. The setting of this attribute was not discussed with @leewujung. 

This new naming scheme differs from the old one because we did not include the `EchoData` group in the names of the variable and dimension. Additionally, we did not provide attributes to the variable. 

Pinging @emiliom to inform him of the new way we will be handling these old times, just in case he takes issue with it.

Below I provide a picture of the new variable name/attributes for a simple EK60 example:

<img width="701" alt="old_time_var_representation" src="https://user-images.githubusercontent.com/53541061/192612521-a0670ceb-0808-44ec-b1b3-c4c48dd68113.png">

b-reyes commented:
Currently all combine tests pass (after simple modifications) except `test_combine_echodata`. The resolution of `test_combine_echodata` seems like it will require further investigation as we are directly comparing Datasets. As discussed with @leewujung, we will momentarily xfail this test and correct it after this PR is done. 

The required simple modifications:

- For each test we need to create a temporary directory and provide the path to the new `combine` method.

- The test `test_combine_attrs` was removed as it was testing an optional input in 
the old combine method that allowed users the ability to combine the attributes in various ways. 

- In `test_combined_encodings` we were just making sure that the time encodings are correct. However, now that we are lazy loading, the encodings relating to dask are included. To make the test pass, I remove the following encodings if they exist: `'chunks', 'preferred_chunks', 'compressor', 'filters'`.

b-reyes commented:
In the first bullet point in the [TODO comment](https://github.com/OSOceanAcoustics/echopype/pull/808#issuecomment-1251690382) above, I reference that we are obtaining an intermitent decompression error. After further investigation, it appears that it is caused by competing processes attempting to write to the same chunk in the final combined zarr store. It should be noted that data corruption will occur. @leewujung has suggested (and I have agreed) that we first complete the structure of `combine_echodata` in this PR and then fix the competing processes issue in a different PR.  

Addtionally, any further integration tests and the resolution of `test_combine_echodata` will take place in a different PR.

leewujung commented on echopype/utils/io.py:
```diff
@@ -32,14 +34,59 @@ def get_files_from_dir(folder):
+
+        else:
+
+            # TODO: below is the encoding we were using for netcdf, we need to make
+            #  sure that the encoding is appropriate for all data variables
+            encoding = (
+                {var: compression_settings for var in ds.data_vars}
+                if compression_settings is not None
+                else {}
+            )
```
In addition to making sure the encoding is appropriate, I suggest wrapping the setting of encoding for both zarr and netcdf into a single function, something like `set_storage_encodings(ds: xr.Dataset, compression_settings: dict, engine: str)` for readability.

Alternatively, have a similar function as in the case of zarr, say `set_netcdf_encodings` so the structure is parallel.

leewujung replied:
In reviewing other parts of the code I came across `utils/coding.py::set_encodings()` that also handles some default encoding. Since there was some encoding stuff before your changes in this `io.py` module, it is likely that that function was not comprehensive. I think it will be good to consolidate all the encoding-related functions into `utils/coding.py` so that this type of repetition/confusion do not happen in the future.

b-reyes replied:
>In addition to making sure the encoding is appropriate, I suggest wrapping the setting of encoding for both zarr and netcdf into a single function, something like set_storage_encodings(ds: xr.Dataset, compression_settings: dict, engine: str) for readability.
> Alternatively, have a similar function as in the case of zarr, say set_netcdf_encodings so the structure is parallel.

That's a good idea, I will go ahead and do that too.

>In reviewing other parts of the code I came across utils/coding.py::set_encodings() that also handles some default encoding. Since there was some encoding stuff before your changes in this io.py module, it is likely that that function was not comprehensive. I think it will be good to consolidate all the encoding-related functions into utils/coding.py so that this type of repetition/confusion do not happen in the future.

I agree that we should move all encoding related items to `utils/coding.py`. I will go ahead and do that. 

From my understanding, `utils/coding.py::set_encodings()` is really only setting the encoding for the time coordinates (i.e. the values in `DEFAULT_TIME_ENCODING`). No other types of encoding are set. Perhaps in a new issue we should consider renaming this function.

leewujung replied:
> From my understanding, utils/coding.py::set_encodings() is really only setting the encoding for the time coordinates (i.e. the values in DEFAULT_TIME_ENCODING). No other types of encoding are set. Perhaps in a new issue we should consider renaming this function.

You are right. Please change the time as you see fit.

b-reyes replied:
Ok, I will create an issue for this and we can do a name change at a later time.

leewujung commented on echopype/convert/set_groups_ek60.py:
```diff
@@ -268,7 +268,6 @@ def set_platform(self, NMEA_only=False) -> xr.Dataset:
         )
-        ds = ds.chunk({"time1": DEFAULT_CHUNK_SIZE["ping_time"]})
 
         if not NMEA_only:
```
I think you can remove this check for `NMEA_only` and the input argument -- this should be remnant of that datagram selection mechanism that was removed in v0.6.2 or right after (i.e. will be part of this release v0.6.3). Please double check but my quick search did not turn up with any usage of this variable.

b-reyes replied:
I double checked and I agree, I do not think this is ever used and is a remnant of the selection mechanism. I will go ahead and remove it.

leewujung commented on echopype/qc/api.py:
```diff
@@ -55,7 +55,8 @@ def coerce_increasing_time(
     """
 
-    ds[time_name] = _clean_ping_time(ds[time_name].values, local_win_len=local_win_len)
+    ping_time_new = _clean_ping_time(ds[time_name].values, local_win_len=local_win_len)
+    ds[time_name].values[:] = ping_time_new
```
Why does the operation need to be broken into 2 pieces?

b-reyes replied:
Are you suggesting that we do the following? If so, I do not have a problem with that. 

`ds[time_name].values[:] = _clean_ping_time(ds[time_name].values, local_win_len=local_win_len)`

leewujung replied:
I was just wondering why the previous one before the change doesn't work. And the oneliner you put looks good.

b-reyes replied:
Ah, I see. The function `_clean_ping_time` returns a numpy array and in the previous version we assign this to the `ds` variable `time_name`. For this reason, we lose all attributes/encodings (if they exist). What we really want to do is just replace the values. 

> And the oneliner you put looks good.

Ok, I will change that into a oneliner.

leewujung replied:
Thanks for the explanation! I completely missed the attributes side of things.

leewujung commented on echopype/test_data/README.md:
```diff
@@ -11,8 +11,6 @@ Most of these files are stored on Git LFS but the ones that aren't (due to file
 - 2019118 group2survey-D20191214-T081342.raw: Contains 6 channels but only 2 of those channels collect ping data
 - D20200528-T125932.raw: Data collected from WBT mini (instead of WBT), from @emlynjdavies
 - Green2.Survey2.FM.short.slow.-D20191004-T211557.raw: Contains 2-in-1 transducer, from @FletcherFT (reduced from 104.9 MB to 765 KB in test data updates)
-- raw4-D20220514-T172704.raw: Contains RAW4 datagram, 1 channel only, from @cornejotux
-- D20210330-T123857.raw: do not contain filter coefficients
```
Please add these 2 lines back so that we remember them. We'll do a thorough review of the files in the upcoming test overhaul.

leewujung commented on echopype/test_data/README.md:
```diff
@@ -24,7 +22,6 @@ Most of these files are stored on Git LFS but the ones that aren't (due to file
 - Winter2017-D20170115-T150122.raw: Contains a change of recording length in the middle of the file
 - 2015843-D20151023-T190636.raw: Not used in tests but contains ranges are not constant across ping times
 - SH1701_consecutive_files_w_range_change: Not used in tests. [Folder](https://drive.google.com/drive/u/1/folders/1PaDtL-xnG5EK3N3P1kGlXa5ub16Yic0f) on shared drive that contains sequential files with ranges that are not constant across ping times.
-- NBP_B050N-D20180118-T090228.raw: split-beam setup without angle data
```
Same as the above, please add this line back

leewujung commented on echopype/utils/io.py:
```diff
@@ -32,14 +34,59 @@ def get_files_from_dir(folder):
+
+
 def save_file(ds, path, mode, engine, group=None, compression_settings=None):
     """Saves a dataset to netcdf or zarr depending on the engine
     If ``compression_settings`` are set, compress all variables with those settings"""
```
```suggestion
    """
    Saves a dataset to netcdf or zarr depending on the engine
    If ``compression_settings`` are set, compress all variables with those settings
    """
```
Just so that the docstring format is uniform...

leewujung commented on echopype/echodata/combine.py:
```diff
@@ -1,19 +1,67 @@
+
 def check_echodatas_input(echodatas: List[EchoData]) -> Tuple[str, List[str]]:
     """
     Ensures that the input ``echodatas`` for ``combine_echodata``
```
```suggestion
    Ensures that the input list of ``EchoData`` objects for ``combine_echodata``
```

leewujung commented on echopype/echodata/combine.py:
```diff
@@ -79,385 +142,270 @@ def check_and_correct_reversed_time(
-    Creates an in-memory (i.e. in RAM) combined ``EchoData``
-    object from the values in ``echodatas``.
+        if group != "Platform/NMEA":
+            # Platform/NMEA is skipped because we found that the times correspond to other
+            # messages besides GPS. This causes multiple times to be out of order and
+            # correcting them is not possible with the current implementation of
+            # _clean_ping_time in qc.api
```
```suggestion
            # Platform/NMEA is skipped because we found that the times correspond to other
            # non-GPS messages are often out of order and correcting them is not possible 
            # with the current implementation of _clean_ping_time in qc.api due to excessive recursion.
            # There is also no obvious advantage in correcting the order of these timestamps.
```

leewujung commented on echopype/echodata/combine.py:
```diff
@@ -79,385 +142,270 @@ def check_and_correct_reversed_time(
+    ValueError
+        If any ``EchoData`` object's ``sonar_model`` is ``None``
+    ValueError
+        If and ``EchoData`` object does not have a file path
```
```suggestion
        If any ``EchoData`` object does not have a file path
```

leewujung commented on echopype/echodata/combine.py:
```diff
@@ -79,385 +142,270 @@ def check_and_correct_reversed_time(
+        than the first time value of the subsequent corresponding
+        ``EchoData`` group, with respect to the order in ``echodatas``
+    RuntimeError
+        If each corresponding ``EchoData`` group in ``echodatas`` do not
```
```suggestion
        If the same ``EchoData`` groups in ``echodatas`` do not
```

leewujung commented on echopype/echodata/combine.py:
```diff
@@ -79,385 +142,270 @@ def check_and_correct_reversed_time(
-    * Attributes from all groups before the combination will be stored in the provenance group,
-      although this behavior may change in future versions.
+    * ``EchoData`` objects are combined by appending their groups individually to a zarr store.
+    * All attributes (besides array attributes) from all groups before the combination will be
+      stored in the ``Provenance`` group.
```
Not clear what "array attributes" are. Do you mean attributes of data variables within a given group?

b-reyes replied:
By "array attributes" I mean group attributes whose value is an array. Perhaps the following is a better statement: 

`All attributes (besides attributes whose values are arrays) from all groups before the combination will be
      stored in the ``Provenance`` group.`

leewujung replied:
The revised statement looks good!

leewujung commented on echopype/echodata/combine.py:
```diff
@@ -79,385 +142,270 @@ def check_and_correct_reversed_time(
+    * ``EchoData`` objects are combined by appending their groups individually to a zarr store.
+    * All attributes (besides array attributes) from all groups before the combination will be
+      stored in the ``Provenance`` group.
     * The ``source_file`` and ``converted_raw_path`` attributes will be copied from the first
```
```suggestion
    * The instance attributes ``source_file`` and ``converted_raw_path`` of the combined ``EchoData`` object will be copied from the first 
```

leewujung commented on echopype/echodata/combine.py:
```diff
@@ -79,385 +142,270 @@ def check_and_correct_reversed_time(
-      ``EchoData`` object in the given list, but this may change in future versions.
+      ``EchoData`` object in the given list.
+    * If any time coordinate in a final combined group is not in ascending order, then it will
+      be corrected according to `#297 <https://github.com/OSOceanAcoustics/echopype/pull/297>`_.
```
Referencing to a very old PR is probably not a good idea in the docstring. I am marking this as a TODO for later.
```suggestion
      be corrected according to `#297 <https://github.com/OSOceanAcoustics/echopype/pull/297>`_.
      # TODO: change this to description of the gist of the employed correction.
```

leewujung commented on echopype/echodata/combine.py:
```diff
@@ -79,385 +142,270 @@ def check_and_correct_reversed_time(
+      be corrected according to `#297 <https://github.com/OSOceanAcoustics/echopype/pull/297>`_.
+      Additionally, the uncorrected time coordinate will be stored in the ``Provenace`` group as
+      a variable and the ``Provenance`` attribute ``reversed_ping_times`` will be set to ``1``.
+    * If no ``zarr_path`` is provided, it will be set to
+      ``'temp_echopype_output/combined_echodatas.zarr'`` in the current working directory.
```
```suggestion
    * If no ``zarr_path`` is provided, the combined zarr file will be
      ``'temp_echopype_output/combined_echodatas.zarr'`` in the current working directory.
```

leewujung commented on echopype/echodata/combine.py:
```diff
@@ -79,385 +142,270 @@ def check_and_correct_reversed_time(
-    # all attributes before combination
-    # { group1: [echodata1 attrs, echodata2 attrs, ...], ... }
-    old_attrs: Dict[str, List[Dict[str, Any]]] = dict()
+    zarr_path = check_zarr_path(zarr_path, storage_options)
```
```suggestion
    # Check the provided zarr_path is valid, or create a temp zarr_path if not provided
    zarr_path = check_zarr_path(zarr_path, storage_options)
```

leewujung commented on echopype/echodata/combine.py:
```diff
@@ -79,385 +142,270 @@ def check_and_correct_reversed_time(
-        raise NotImplementedError(
-            "Lazy representation of combined EchoData object has not been implemented yet."
-        )
+    sonar_model, echodata_filenames = check_echodatas_input(echodatas)
```
```suggestion
    # Ensure the list of all EchoData objects to be combined are valid
    sonar_model, echodata_filenames = check_echodatas_input(echodatas)
```

leewujung reviewed (commented):
@b-reyes : I've reviewed all changes except for the new `zarr_combine.py` module -- will save that for tomorrow morning when I have a clear head.

I added many textual changes to the docstrings and a few embedded comments in the code for readability.

One question:
Do you actually mean to have `tests/echodata/test_cluster_dump/test_zarr_combine.yaml` in the commits?

b-reyes commented:
> @b-reyes : I've reviewed all changes except for the new `zarr_combine.py` module -- will save that for tomorrow morning when I have a clear head.

Sounds good, I will most likely wait to address these until you review all of the code and focus my effort on the data corruption issue. 

> One question: Do you actually mean to have `tests/echodata/test_cluster_dump/test_zarr_combine.yaml` in the commits?

I am not sure where this came from. Perhaps it was generated when I was testing out creating a `@gen_cluster` style of test with dask. I will look into this and  remove it, if it is not necessary.

leewujung commented on echopype/echodata/zarr_combine.py:
```diff
@@ -0,0 +1,742 @@
+                times = ds[time].values
+                if isinstance(times, np.ndarray):
+                    # store first time if we have an array
+                    first_times.append(times[0])
```
I think checking for the first non-NaT times and not just the first element would be more general. Do you see a problem with doing so?

b-reyes replied:
I have come across the case where the `Platform` group's `time1` variable only has a single `NaT` value (I think this is manually set by us). In this case, checking for the first non-NaT time would not work. That's really the only issue I can see.

leewujung replied:
Ah I see, ok, then we can just do the first element then.

leewujung commented on echopype/echodata/zarr_combine.py:
```diff
@@ -0,0 +1,742 @@
+        # The sonar_model for the new combined EchoData object
+        self.sonar_model = None
+
+    def _check_ascending_ds_times(self, ds_list: List[xr.Dataset], ed_name: str) -> None:
```
Thinking to avoid confusing it'd be good to call this `_check_ascending_ds_sequence` or something similar so that it is not confused with `orchestrate_reverse_time_check` that actually checked all the timestamps.

b-reyes replied:
In my mind reverse and ascending are two different properties. For example, the times can be ascending amongst the datasets, but still have reversed times in them. 

`_check_ascending_ds_sequence` is a little to general in my opinion, since we are specifically checking only the times.

I cannot think of a better name, but I am open to other suggestions.

leewujung replied:
Sure I am fine with the current time.

leewujung commented on echopype/echodata/zarr_combine.py:
```diff
@@ -0,0 +1,742 @@
+
+        Notes
+        -----
+        This method creates the following class variables:
+        dims_df: pd.DataFrame
+            Dataframe with column as dim names, rows as the
+            different Datasets, and values as the length of the dimension
+        dims_sum: dict
+            Keys as the dimension name and values as the corresponding
+            sum of the lengths across all Datasets
+        dims_csum: dict
+            Keys as the dimension name and values as a dictionary of
+            the corresponding cumulative sum of the lengths across
+            all Datasets
+        dims_max: dict
+            Keys as the dimension name and values as the corresponding
+            maximum length across all Datasets
```
I wonder if it is worthwhile to initialize these as None in `__init__`, just for readability. The docstring here is very useful, and I wonder these should be also -- actually, I always have the question of where docstring of class/instance attributes should go...

b-reyes replied:
> I wonder if it is worthwhile to initialize these as None in __init__, just for readability.

This is a good point, I think that is good coding practice. I will go ahead and add these to `__init__`

> The docstring here is very useful, and I wonder where these should be also -- actually, I always have the question of where docstring of class/instance attributes should go...

Yes, I am always unsure of where to put these types of docstring, so I usually just put them in the method that creates them. Should we leave them where they are right now?

leewujung replied:
Right, maybe we can leave what you have here in this method since it is where things are actually happening.

Maybe in the `__init__` section you could add a 1 line comment to send people to this method to find the details?

b-reyes replied:
Sounds good, I will put a comment pointing to `_get_ds_info` in `__init__`.

leewujung commented on echopype/echodata/zarr_combine.py:
```diff
@@ -0,0 +1,742 @@
+
+        return temp_arr, chnk_shape
+
+    def _set_encodings(
+        self, encodings: Dict[str, dict], name: Hashable, val: xr.Variable, chnk_shape: list
+    ) -> None:
```
I'm trying to wrap my head around this and `utils/io.py::set_zarr_encodings` that only gets used in `utils/io.py::save_file`, which is called when groups are saved to file using `EchoData.to_zarr/to_netcdf`. It seems the operations are partially overlapping so encodings for some variables are overwritten (maybe with the same settings?) when the data is actually written to disk using the `to_zarr/to_netcdf` functions?

b-reyes replied:
In this function I am gathering the encodings (except the lazy encodings) of each variable so that I can pass them off to the delayed `to_zarr` function. By supplying the encodings, we can construct all of the metadata for the final combined EchoData object. Here we do not want to use the encodings such as `chunks` because this encoding would correspond to the encoding of the first dataset in the datasets that are to be combined. Instead, we want to set chunks to be uniform (this is done at the end of the function).

> It seems the operations are partially overlapping so encodings for some variables are overwritten (with the same settings?)

We are not overwriting any encodings, just collecting them. The statement that sets the compressor is only needed if no compressor is provided. No compressor is present if the provided variable was obtained from an in-memory EchoData group i.e. we are combining in-memory EchoData objects.

leewujung replied:
Ah now I see that the collection only happens here. Just to clarify: by "lazy encodings" do you mean like `chunk`?

leewujung replied:
Also, after looking at the use of this function some more, I also realized why I was confused when first reading it. I think I was not expecting `encodings` in `_construct_lazy_ds_and_var_info` in modified in place with `self._set_encodings(encodings, name, val, chnk_shape)`. 

How about making this function to return a new dict (let's call it `new_dict` for now), and use `encodings.update(new_dict)` to add the new encoding key-value pair to the dict `encodings` in `_construct_lazy_ds_and_var_info`?

b-reyes replied:
> Ah now I see that the collection only happens here. Just to clarify: by "lazy encodings" do you mean like chunk?

Here, "lazy encodings" are `chunks` and `preferred_chunks`. 

I will address the `set_encodings` comment in the other more in-depth comment you make below.

leewujung replied:
Thanks for the explanation!

leewujung commented on echopype/echodata/zarr_combine.py:
```diff
@@ -0,0 +1,742 @@
+
+    def _get_temp_arr(self, dims: List[str], dtype: type) -> Tuple[type(dask.array), list]:
+        """
+        Constructs a temporary (or dummy) array representing a
+        variable in its final combined form.
```
might be useful to add something about the array has the shape of the final combined data

b-reyes replied:
Sounds good, I will add another sentence stating this.

leewujung reviewed (approved):
Hey @b-reyes : I am done reviewing `combine_zarr` -- it's cool to see the operations neatly factored out!  My inline comments are mostly cosmetic, except for the questions related to the various set encodings functions. I think the majority of the infrastructure you have set up will just work for whatever changes in the distributed writing you end up implementing to resolve the data corruption problem. Also, it seems that it would 2 lines of changes in `._append_ds_list_to_zarr()` to write the zarr stores sequentially? Anyways, since the resolution of data corruption is in the works, it doesn't matter here.

I'm marking this as "approved" so that once you think you've address the comments I made or decide to push some of those into your next PR, feel free to merge this. Thanks!

b-reyes commented on echopype/echodata/combine.py:
```diff
@@ -79,385 +142,274 @@ def check_and_correct_reversed_time(
+      ``EchoData`` object will be copied from the first ``EchoData`` object in the given list.
+    * If any time coordinate in a final combined group is not in ascending order, then it will
+      be corrected according to `#297 <https://github.com/OSOceanAcoustics/echopype/pull/297>`_.
+      # TODO: change this to description of the gist of the employed correction.
```
@leewujung I accepted and committed this change, but I didn't realize that it was in the middle of a docstring. I think we should place this somewhere else because it will show up in the generated docs for users.

leewujung replied:
Yes, I agree, perhaps move it to immediately below the entire """ block so we would remember (I'll linked this comment in #798). Looking at this again I am thinking if we have a section in the docs that explain how we do the correction, we can put that link into this docstring here.

b-reyes replied:
> Looking at this again I am thinking if we have a section in the docs that explain how we do the correction, we can put that link into this docstring here.

Yes, that is the best solution.

b-reyes commented:
@leewujung I have addressed or responded to all of your comments. Please give this another review.

leewujung commented on echopype/echodata/zarr_combine.py:
```diff
@@ -0,0 +1,742 @@
+                # collect the names of all constant variables/dimensions
+                const_names.append(str(name))
+
+            elif name not in ds_model.dims:
+
+                # create lazy DataArray representations corresponding to the variables
+                temp_arr, chnk_shape = self._get_temp_arr(list(val.dims), val.dtype)
+                xr_vars_dict[name] = (val.dims, temp_arr, val.attrs)
+
+                self._set_encodings(encodings, name, val, chnk_shape)
+
+            elif name in self.append_dims:
+
+                # create lazy DataArray for those coordinates that can be appended to
+                temp_arr, chnk_shape = self._get_temp_arr(list(val.dims), val.dtype)
+                xr_coords_dict[name] = (val.dims, temp_arr, val.attrs)
+
+                self._set_encodings(encodings, name, val, chnk_shape)
```
Just to make sure I convey what I tried to describe [above under `_set_encodings()`](https://github.com/OSOceanAcoustics/echopype/pull/808/files/4330422c8f90c9f05f9fafba9e0bb709ea9e55fe#r984185663), I am suggesting to have self._set_encodings to return a new dict and add that to `encodings` here, like:
```python
# add new encodings to the everything-in-one encodings dict here
encodings.update(self._set_encodings(name, val, chnk_shape)) 
```

leewujung replied:
Also, just because the first and last line of both of these `elif` are identical, i think the logic here can be:
```python
        for name, val in ds_model.variables.items():

            # get all dimensions of val that are also append dimensions
            append_dims_in_val = set(val.dims).intersection(self.append_dims)

            if not append_dims_in_val:

                # collect the names of all constant variables/dimensions
                const_names.append(str(name))

            else:

                # create lazy DataArray representations corresponding to the variables
                temp_arr, chnk_shape = self._get_temp_arr(list(val.dims), val.dtype)

                if name not in ds_model.dims:
                     xr_vars_dict[name] = (val.dims, temp_arr, val.attrs)
                elif name in self.append_dims:
                     xr_coords_dict[name] = (val.dims, temp_arr, val.attrs)
                
                # [below i just swapped in the suggestion in the comment immediately above]
                # add new encodings to the everything-in-one encodings dict here
                encodings.update(self._set_encodings(name, val, chnk_shape)) 
```

b-reyes replied:
Thank you for the more in depth comment. I agree with your suggestions. I will implement them.

leewujung reviewed (approved):
@b-reyes : thanks for the changes and comments. I am done with reviewing everything now!

b-reyes commented:
@leewujung I am done addressing your comments from the second round of revision. The only major change that was made is now `_set_encodings` returns a dictionary, the `elif` block logic was modified, and I change `_set_encodings` to `_get_encodings` (that seemed a little more appropriate with the changes). This is now ready for a third round of revision.

leewujung commented:
Thanks @b-reyes : everything looks good here. 
One last thing: could you please add a quick explanation what the lazy encodings in the docstring like what you said in the comment above? Just for future reference.

> "lazy encodings" are chunks and preferred_chunks.

Feel free to merge this afterwards. Awesome to see these changes in!

b-reyes commented:
> One last thing: could you please add a quick explanation what the lazy encodings in the docstring like what you said in the comment above? Just for future reference.

@leewujung I have added a more specific portion to the `_get_encodings` docstring as we discussed. Once all the tests have passed, I will merge.
